### PR TITLE
dockerTools: use skopeo on the right platform

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
###### Motivation for this change

When building a Docker image using any one of the `dockerTools` in `pkgsCross`, a `skopeo` on the host platform is chosen and usually this leads to unrecognised Exec format. This PR fixes this issue by assigning the right build packages on the build platform.
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by arianvp** [comment]
Lets just get rid of the `pkgs` argument altogether. To make sure we're not making more mistakes like this.

**Comment by dingxiangfei2009** [comment]
@arianvp  Good point. I am going to try out, and maybe also fix, the `examples`, too.

**Comment by dingxiangfei2009** [comment]
@arianvp `examples` are still solid.

**Comment by arianvp** [comment]
Great. LGTM!

**Comment by arianvp** [comment]
@GrahamcOfBorg test docker-tools

**Comment by arianvp** [comment]
@GrahamcOfBorg test docker-tools

**Comment by arianvp** [comment]
Alas, the error doesn't seem to be transient on `x86_64` https://logs.nix.ci/?key=nixos/nixpkgs.63865&attempt_id=91cb4421-6100-4582-9c07-25d92ee2f597    Not sure what is going on here.

**Comment by dingxiangfei2009** [comment]
@arianvp I notice that on `master` (aka. https://github.com/NixOS/nixpkgs/commit/f3282c8d1e0ce6ba5d9f6aeddcfad51d879c7a4a#diff-f5fad34a5e280fe300df690ac61dc673), `dockerTools.examples.nginx` cannot build already. I am doing a bisect.

**Comment by dingxiangfei2009** [comment]
@arianvp This is the reason for that error: https://github.com/NixOS/nixpkgs/commit/4a4561ce244c0cea1cb07fd02f176b11f094f570#diff-f5fad34a5e280fe300df690ac61dc673R122

It is deleting the VM drive before `qemu` exits, which makes it panic.

**Comment by dingxiangfei2009** [comment]
I don't know if I have access, but

@GrahamcOfBorg test docker-tools

**Comment by arianvp** [comment]
@GrahamcOfBorg test docker-tools

**Comment by dingxiangfei2009** [comment]
@GrahamcOfBorg test docker-tools

**Comment by arianvp** [comment]
I think I just spawned a test seconds before you force-pushed. So it might fail again haha

**Comment by arianvp** [comment]
@GrahamcOfBorg test docker-tools

**Comment by arianvp** [comment]
Hmm, the test seems to time out on x86_64 whilst `docker load`'ing the container. that doesn't sound good :/

**Comment by dingxiangfei2009** [comment]
@arianvp I tried the `docker load` command locally with the `x86_64` image, and it works fine.

I want to try running these tests locally. How can I run this specific test?

**Comment by arianvp** [comment]
`nix-build ./nixos/tests/docker-tools.nix`

**Comment by dingxiangfei2009** [comment]
@arianvp Work fine here. I cannot reproduce the timeout.

See the gist:
https://gist.github.com/dingxiangfei2009/91b7d763853f4c6823e310792adf420f

**Comment by arianvp** [comment]
@dingxiangfei2009 the reason is that not all build machines support nested kvm, so they emulate x86 and are slow.   so this is why it's failing. Not because there is something wrong with the test.

Lets get this merged

**Comment by srhb** [comment]
This looks correct to me and I don't think there should be a reason to suspect that it should be slower on Hydra than the current release-blocker docker test. Let's keep an eye on it regardless. :)

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -250,6 +253,12 @@ rec {
 
       ${postUmount}
       '');
+    in
+    runCommand name {} ''
+      mkdir -p $out
+      cd ${result}
+      cp layer.tar json VERSION $out
+    '';
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:261]: Could you clarify the purpose of this `runCommand` wrapper?
It was found to be broken: https://github.com/NixOS/nixpkgs/pull/80068/files#r613275987

Comment (roberth) [reply pkgs/build-support/docker/default.nix:261]: If the goal is to reduce output size, it's generally better to do it in the original derivation. Doing so reduces the need for garbage collection, reduces I/O and reduces the storage required in cases where everything is cached eagerly.

Comment (ajs124) [reply pkgs/build-support/docker/default.nix:261]: How does this reduce output size, if I just want to extract some files from a container image? I'll need to create a tar, just to unpack it again in the next build step.

Comment (roberth) [reply pkgs/build-support/docker/default.nix:261]: It seems to have very little utililty. The only role I can think of for this `runCommand` call is to keep redundant files from propagating to other derivations, but that should probably be done in `result` itself instead.

@srhb is there anything to this code that we're not thinking of?

Comment (roberth) [reply pkgs/build-support/docker/default.nix:261]: I'm fairly confident in my guess that it was to remove the files created by `createEmptyDiskImage`. I've made it write to a different location so we don't need this extra derivation anymore. Done in https://github.com/NixOS/nixpkgs/pull/80068

Comment (roberth) [reply pkgs/build-support/docker/default.nix:261]: I realize now that I asked the wrong person about these lines. Should've been @dingxiangfei2009 
Oh well.

**Pull request merged**